README.md
# Swerve
_Quick and easy multithreaded file server for local development_

[![Build Status](https://travis-ci.org/Commander-lol/rust-swerve.svg?branch=master)](https://travis-ci.org/Commander-lol/rust-swerve) [![License](https://img.shields.io/crates/l/swerve.svg)](https://crates.io/crates/swerve) [![Version Number](https://img.shields.io/crates/v/swerve.svg)](https://crates.io/crates/swerve)

## About
A common misconception about swerve is that the name is a play on the word "serve", as in the purpose of a file server.
Contrary to this, swerve actually stands for "Simple Werve" - the goal being to make it super simple to werve your 
files from anywhere on the filesystem.

While swerve does nothing innovative, the goal is to have a fast performing file server with minimal set up and
tear down time, as well as minimal request time spent inside the server itself. Current features are limited to
serving files, but the roadmap includes API mocking to make front end development even easier.

## Installation

### Prebuilt
1. Download the binary for your OS from the [releases](https://github.com/Commander-lol/rust-swerve/releases) page
	- `swerve.x86_64-unknown-linux-gnu` = `Generic Linux`
	- `swerve.x86_64-apple-darwin` = `macOS`
3. Rename the binary to `swerve`
2. Put `swerve` somewhere in your `$PATH`, or modify your `$PATH` to inlcude the folder containing `swerve`
3. Do one of two things:
	1. Open up a command prompt and `cd` to the folder containing your files and run `swerve`
		```bash
		cd "$HOME/projects/my_awesome_site"
		swerve
		```
	2. Open up a command prompt and run `swerve`, specifying your target directory
		```bash
		swerve -d "$HOME/projects/my_awesome_site"
		```
		
### From source
To build and install swerve from source, you will need `rustc 1.27.0-nightly` or later as well as the Cargo package 
manager

1. In a command prompt, run `cargo install swerve`

This will build a copy of swerve on your local machine in case there isn't a binary built for your platform.
A little bit anticlimactic, but pretty simple.

### Usage
You can print out the usage text at any time by running `swerve -h` or `swerve --help`. The currently supported options
are:

Option | Param | Description | Notes
-------|-------|-------------|--------
`-h`, `--help` | | Print out the help text | 
`-d`, `--dir` | path; `string` | Use the target directory as the root for serving files | File paths are jailed, and will not be able to escape the root directory. Defaults to the directory in which the `swerve` command was run
`-p`, `--port` | port; `unsigned int` | Run the server listening on the given port | Defaults to `8200`
`-a`, `--address` | address; `string` | Bind the server to this network address, allowing remote connections | Defaults to `localhost`, preventing remote connections
`-t`, `--threads` | thread count; `unsigned int` | Create this number of threads in the thread pool, to use for serving files concurrently | Defaults to `32`. Performance may be improved on lower end machines by decreasing the number of worker threads
`-q`, `--quiet` | | Don't print anything to stdout | Useful if you're spawning swerve from another process and need to monitor stdout
`--no-index` | | Don't attempt to serve an `index.html` file from a directory path | By default, swerve will treat directories as requests for an `index.html` file. e.g. `/foo/bar` is treated as `/foo/bar/index.html`
`-u`, `--upload` | | Accept form uploads via a `POST` request to the `/uploads` endpoint | The `file_handling` and `field_handling` config value affect how forms are handled. Setting them to `Log` or `File` will log the values or write them to disk respectively
`-U`, `--upload-path` | path; `string` | Accept form uploads via a `POST` request to the specified endpoint | See the `--upload` option for information on configuration
`--license` | | Print out the GPL 3 license text | GPL compliance is cool, kids


CHANGELOG.md
# Changelog
All notable changes to this project will be documented in this file.

The format is based on [Keep a Changelog](http://keepachangelog.com/en/1.0.0/)
and this project adheres to [Semantic Versioning](http://semver.org/spec/v2.0.0.html).

## [Unreleased]

### Changed
- Upgrade rocket dependency to `0.3.16`

## [0.3.0]

### Added
- [DEV] Initial test scaffolding
- Lua scripting support as a route handler in the `config.yml` file

## [0.2.0]

### Added
- Support for `-u, --uploads` to add a path that accepts form submissions
- Support for -U=<path>, --upload-path=<path>` to add a custom path that accepts form submissions
- `.swerve` configuration directory support. Swerve will look for a `.swerve` directory in the target directory, containing a `config.yml` file that can define various server options, including command line options
- Support "Log" and "File" handlers for files in form submissions
- Support "Log" handler for fields in form submissions
- [UNSTABLE] Initial support for custom route scripting

### Fixed
- Incorrect file type returned for `.wasm` files

## [0.1.1]

### Fixed
- [DEV] Minor errors in metadata files

## [0.1.0]

### Added
- Serve files from the current directory
- Serve files from the target directory
- Serve index files from directories
- Configure number of worker threads
- Bind to custom address and port


src/cli/cli.rs
pub use std::path::PathBuf;
pub use std::env::current_dir;
pub use std::default::Default;

pub const USAGE: &'static str = "
Static file swerver and api mocker for local development. 
For full documentation, visit https://swerve.louiscap.io.

Usage:
    swerve [options]

General Options:
    -h, --help                       Display this text
    -q, --quiet                      Don't print anything to stdout
    --license                        Show the GPL v3 license text

Web Server Options:
    --no-index                       Don't serve index.html files for directory paths
    -d=<path>, --dir=<path>          Root directory; defaults to cwd
    -p=<port>, --port=<port>         Port; defaults to 8200
    -a=<addr>, --address=<addr>      Address to bind to (e.g. 10.0.0.15); defaults to localhost
    -c=<path>, --config=<path>       Path to the .swerve config file
    -t=<num>, --threads=<num>        Number of worker threads to use for serving files; defaults to 32

Data Handling Options:
    -u, --upload                     Support file uploads to '/upload'
    -U=<path>, --upload-path=<path>  Set the url path that will accept file uploads. Implies 'upload' flag if not present

Swerve Copyright (C) 2018  Louis Capitanchik
Licensed under GPLv3+
";

#[derive(Debug, Deserialize, Clone)]
pub struct Args {
    pub flag_dir: Option<String>,
    pub flag_port: Option<u16>,
    pub flag_config: Option<String>,
    pub flag_threads: Option<u16>,
    pub flag_address: Option<String>,
    pub flag_help: bool,
    pub flag_quiet: bool,
    pub flag_no_index: bool,
	pub flag_upload: bool,
	pub flag_upload_path: Option<String>,
	pub flag_license: bool,
}

impl Args {
    pub fn get_dir(&self) -> PathBuf {
        let dir_flag = self.flag_dir.clone();
        PathBuf::from(
            dir_flag.unwrap_or(
            current_dir().unwrap_or(
            PathBuf::from("")
            ).to_string_lossy().into_owned())
        )
    }
}

impl Default for Args {
	fn default() -> Self {
		Args {
			flag_dir: Some(String::from("")),
			flag_port: Some(8000),
			flag_config: None,
			flag_threads: Some(32),
			flag_address: Some(String::from("localhost")),
			flag_help: false,
			flag_quiet: false,
			flag_no_index: false,
			flag_upload: false,
			flag_upload_path: None,
			flag_license: false,
		}
	}
}